src/Orleans.Transactions/State/NamedTransactionalStateStorageFactory.cs

```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Runtime;
+using Orleans.Transactions.Abstractions;
+using Orleans.Storage;
+
+namespace Orleans.Transactions
+{
+    public class NamedTransactionalStateStorageFactory : INamedTransactionalStateStorageFactory
+    {
+        private readonly IGrainActivationContext context;
+
+        public NamedTransactionalStateStorageFactory(IGrainActivationContext context)
+        {
+            this.context = context;
+        }
+        public ITransactionalStateStorage<TState> Create<TState>(string storageName)
+            where TState : class, new()
+        {
+            // Try to get ITransactionalStateStorage from factory
+            ITransactionalStateStorageFactory factory = string.IsNullOrEmpty(storageName)
+                ? this.context.ActivationServices.GetService<ITransactionalStateStorageFactory>()
+                : this.context.ActivationServices.GetServiceByName<ITransactionalStateStorageFactory>(storageName);
+            if (factory != null) return factory.Create<TState>();
+
+            // Else try to get storage provider and wrap it
+            IStorageProvider storageProvider = string.IsNullOrEmpty(storageName)
+                ? this.context.ActivationServices.GetService<IStorageProvider>()
+                : this.context.ActivationServices.GetServiceByName<IStorageProvider>(storageName);
+            if (storageProvider != null) return new TransactionalStateStorageProviderWrapper<TState>(storageProvider, context);
+
+            throw new InvalidOperationException($"Transactional storage {storageName} was not found.");
```

Comment:
We could flesh this error message out more, since it's an important one. Something like:
`No implementation of either {nameof(ITransactionalStateStorageFactory)} or {nameof(IStorageProvider)} was found while attempting to create transactional state storage.`
or `No implementation of either (a) or (b) with the name (name) was found...`
depending on whether or not `storageName` is set.

If might make sense to if/else the entire method based on a single check for `string.IsNullOrEmpty(storageName)` in that case.

Reply:
Agree that the error could be more fleshed out.  Will do.

As for the storage name being unset, that is valid.  In those cases it should use the default, it will only reach this condition if the name is set and no storage of that name exists, or name is unset and no default exists for either the storage or storage providers.